exporter/bmchelixexporter/metrics_producer.go

```diff
@@ -0,0 +1,260 @@
+	// Extract parent entity information
+	parentEntityTypeID := fmt.Sprintf("%s_container", helixMetric.Labels["entityTypeId"])
+	parentEntityID := fmt.Sprintf("%s:%s:%s:%s", helixMetric.Labels["source"], helixMetric.Labels["hostname"], parentEntityTypeID, parentEntityTypeID)
+
+	// Create a parent entity if not already created
+	if _, exists := containerParentEntities[parentEntityID]; !exists {
+		parentMetric := BmcHelixMetric{
+			Labels: map[string]string{
+				"entityId":               parentEntityID,
+				"entityName":             parentEntityTypeID,
+				"entityTypeId":           parentEntityTypeID,
+				"hostname":               helixMetric.Labels["hostname"],
+				"source":                 helixMetric.Labels["source"],
+				"isDeviceMappingEnabled": helixMetric.Labels["isDeviceMappingEnabled"],
+				"hostType":               helixMetric.Labels["hostType"],
+				"metricName":             "identity", // Represents the parent entity itself
+			},
+			Samples: []BmcHelixSample{}, // Parent entities don't have samples
+		}
+		containerParentEntities[parentEntityID] = parentMetric
+		helixMetrics = append(helixMetrics, parentMetric)
+	}
+
+	// Add parent reference to the child metric
+	helixMetric.Labels["parentEntityName"] = parentEntityTypeID
+	helixMetric.Labels["parentEntityTypeId"] = parentEntityTypeID
+
+	return append(helixMetrics, helixMetric)
+}
+
+// createHelixMetric converts a single OpenTelemetry metric into a BmcHelixMetric payload
+func (mp *MetricsProducer) createHelixMetric(metric pmetric.Metric, resourceAttrs map[string]string) (*BmcHelixMetric, error) {
+	labels := make(map[string]string)
+	labels["source"] = "OTEL"
+
+	// Add resource attributes as labels
+	for k, v := range resourceAttrs {
+		labels[k] = v
+	}
+
+	// Set the metric unit
+	labels["unit"] = metric.Unit()
+
+	// Set the host type
+	labels["hostType"] = "server"
+
+	// Indicates the monitor in the hierarchy that is mapped to the device
+	labels["isDeviceMappingEnabled"] = "true"
+
+	// Update the metric name for the BMC Helix payload
+	labels["metricName"] = metric.Name()
+
+	// Samples to hold the metric values
+	samples := []BmcHelixSample{}
+
+	// Handle different types of metrics (sum and gauge)
+	// BMC Helix only supports simple metrics (sum, gauge, etc.) and not histograms or summaries
+	switch metric.Type() {
+	case pmetric.MetricTypeSum:
+		dataPoints := metric.Sum().DataPoints()
+		for i := 0; i < dataPoints.Len(); i++ {
+			samples = mp.processDatapoint(samples, dataPoints.At(i), labels, metric, resourceAttrs)
+		}
+	case pmetric.MetricTypeGauge:
+		dataPoints := metric.Gauge().DataPoints()
+		for i := 0; i < dataPoints.Len(); i++ {
+			samples = mp.processDatapoint(samples, dataPoints.At(i), labels, metric, resourceAttrs)
+		}
+	default:
+		return nil, fmt.Errorf("unsupported metric type %s", metric.Type())
+	}
+
+	// Check if the entityTypeId is set
+	if labels["entityTypeId"] == "" {
```

Comment:
Too bad we can't figure out there is no `entityTypeId` attribute before processing all data points 🤷‍♂️